### PR TITLE
Bitwise Carries protocol

### PR DESCRIPTION
**Pull request by taikiy**
This diff implements `Carries` protocol. "6.3 Computing the carry bits" from the [paper](https://iacr.org/archive/tcc2006/38760286/38760286.pdf), "Unconditionally Secure Constant-Rounds Multi-party Computation for Equality, Comparison, Bits, and Exponentiation".

`Carries` computes... well, the carries. It takes two bitwise shared secrets, and outputs a bitwise share indicating the indice where bit-carries occur.

This by itself is not useful, but this is a part of the bitwise sum protocol, which I will send out the PR next.

### CODE REVIEW

File: src/protocol/boolean/carries.rs [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    ) -> Result<Vec<CarryPropagationShares<F>>, BoxError> {
+        let xor_ctx = ctx.narrow(&Step::AXorB);
+        let or_ctx = ctx.narrow(&Step::AOrB);
+        let (p_futures, k_futures): (Vec<_>, Vec<_>) = zip(a, b)
+            .enumerate()
+            .map(|(i, (&a_bit, &b_bit))| {
+                let c1 = xor_ctx.narrow(&BitOpStep::Step(i));
+                let c2 = or_ctx.narrow(&BitOpStep::Step(i));
+                (
+                    async move { xor(c1, record_id, a_bit, b_bit).await },
```
Comment (benjaminsavage) [inline src/protocol/boolean/carries.rs:120]: I get what you're doing here (this is indeed computing an XOR). But I think that this way you've unfortunately replicated the exact same multiplication from step_1. Given we are already passing in `s` this doesn't require any more communication and can be done locally.

File: src/protocol/boolean/carries.rs [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+                    async move { or(c2, record_id, a_bit, b_bit).await },
```
Comment (benjaminsavage) [inline src/protocol/boolean/carries.rs:121]: Same here. You're doing an extra multiplication this way when this should be an entirely local operation.

File: src/protocol/boolean/carries.rs [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+
+    /// Step 2. for `i=0..l=1`, generate a vector of tuples `([s], [p], [k])` where
+    /// `[p_i] = [a_i] + [b_i] - 2*[s_i]`, and `[k_i] = 1 - [s_i] - [p_i]`
+    ///
+    /// The interpretations is:
+    ///   * `s_i = 1` iff carry is set at position `i` (i.e. `a_i + b_i = 2`)
+    ///   * `p_i = 1` iff carry would be propagated at position `i` (i.e. `a_i + b_i = 1`)
+    ///   * `k_i = 1` iff a carry would be killed at position `i` (i.e. `a_i + b_i = 0`)
+    ///
+    /// Therefore, we compute `p_i = XOR(a_i, b_i)`, and `k_i = !OR(a_i, b_i)`
```
Comment (benjaminsavage) [inline src/protocol/boolean/carries.rs:89]: This comment is now out of date.

File: src/protocol/boolean/carries.rs [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+    /// computes `([f_0],...,[f_l]) ← PRE○([e_0],...,[e_l])`.
+    async fn step3<F: Field>(
+        e: &[CarryPropagationShares<F>],
+        ctx: ProtocolContext<'_, Replicated<F>, F>,
+        record_id: RecordId,
+    ) -> Result<Vec<CarryPropagationShares<F>>, BoxError> {
+        let l = e.len();
+        let futures = (0..l).map(|i| {
+            let c = ctx.narrow(&BitOpStep::Step(i));
+            async move { Self::fan_in_carry_propagation(&e[0..=i], c, record_id).await }
```
Comment (benjaminsavage) [inline src/protocol/boolean/carries.rs:134]: I am 100% confident that this is doing a ton of duplicate work. I'm OK landing it as-is, but we should attempt to optimize it later.

**Review by benjaminsavage — APPROVED** [review]
This protocol is super confusing and seems insanely costly. Let's land this and I'll put up another diff that implements bitwise addition in a simper, more straightforward way. We can compare benchmarks and go with the more efficient approach.

File: src/protocol/boolean/carries.rs [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+        record_id: RecordId,
+    ) -> Result<CarryPropagationShares<F>, BoxError> {
+        let l = e.len();
+
+        // TODO(taikiy): Optimization: Could run the following 2 blocks in parallel
+
+        // Step 1. fan-in AND. for `i=1..l, ∧ [p_i]`
+        let c = ctx.narrow(&Step::FanInAndP);
+        let mut b = e[0].p;
+        for (i, &x) in e.iter().enumerate().skip(1) {
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply src/protocol/boolean/carries.rs:164]: I think that you are looking for async iterator support for that.  The await really makes this harder.  That this is sequential is a bit of a problem though.  Is there a way to parallelize this?

[the remark this answers is not in the document]
Comment (taikiy) [reply src/protocol/boolean/carries.rs:164]: Damgård's paper describes a logic where we can partition this in blocks and run them concurrently. I'll optimize this later in a separate diff.

File: src/protocol/boolean/carries.rs [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+        //TODO(taikiy): These two futures are independent. How can we run them at once?
+        let p = try_join_all(p_futures).await?;
+        let k = try_join_all(k_futures).await?;
+
+        Ok(zip(p, k)
+            .zip(s)
+            .map(|((p_bit, k_bit), &s_bit)| CarryPropagationShares {
+                s: s_bit,
+                p: p_bit,
+                k: Replicated::one(ctx.role()) - k_bit,
```
[the remark this answers is not in the document]
Comment (taikiy) [reply src/protocol/boolean/carries.rs]: I think this code is outdated. This part looks like this now:
```
    let p_bit = a_bit + b_bit - (s_bit * F::from(2));
    CarryPropagationShares {
        s: s_bit,
        p: p_bit,
        k: one - s_bit - p_bit,
    }
```

File: src/protocol/boolean/carries.rs [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+                let c1 = xor_ctx.narrow(&BitOpStep::Step(i));
+                let c2 = or_ctx.narrow(&BitOpStep::Step(i));
+                (
+                    async move { xor(c1, record_id, a_bit, b_bit).await },
+                    async move { or(c2, record_id, a_bit, b_bit).await },
+                )
+            })
+            .unzip();
+
+        //TODO(taikiy): These two futures are independent. How can we run them at once?
```
[the remark this answers is not in the document]
Comment (taikiy) [reply src/protocol/boolean/carries.rs]: This code is already removed, but I'll try this technique the next time I need it :)

**Pull request merged**